### ISSUE TITLE
Does not recognize installed WSL distro

### ISSUE THREAD

**Issue by dmchitnis** [microsoft/vscode-remote-release#3910]
Issue Type: <b>Bug</b>

1. Click on Open a Remote Window
2. Select Remote-WSL: New Window
3. Notification pops up in bottom right that No WSL distros found.

Ubuntu 20.04 distro is installed using Add-AppxPackage as for some reason it won't install from Microsoft Store. I was able to run sudo apt update. Validated that C:\Program Files\Microsoft VS Code\bin is in the path. echo $PATH in Ubuntu also includes the path. which code in Ubuntu responds with /mnt/c/Program Files/Microsoft VS Code/bin/code. Running 'code .' hangs with no errors and does nothing.

Extension version: 0.50.1
VS Code version: Code 1.50.1 (d2e414d9e4239a252d1ab117bd7067f125afd80a, 2020-10-13T15:06:15.712Z)
OS version: Windows_NT x64 10.0.18363

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz (8 x 2112)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>opengl: enabled_on<br>protected_video_decode: enabled<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|15.85GB (5.67GB free)|
|Process Argv|--crash-reporter-id cee9e6f1-b0c6-4251-8ad1-d0c49363375f|
|Screen Reader|no|
|VM|0%|
</details>

**Comment by bamurtaugh** [comment]
Could you try installing another distro from the Microsoft store to verify if it'll work with Remote-WSL?

**Comment by dmchitnis** [comment]
I have tried all versions of ubuntu and debian as well, although not from Microsoft Store, but using Add-AppxPackage in powershell (for some reason Microsoft Store won't download & install). Where does VS Code look for the info about installed distros?

**Comment by PavelSosin-320** [comment]
After the recent Windows update, I found that all wsl distros have begun to start in a different way depends on the method which is used to start them:
1. Ubuntu 20.04 from Application Ubuntu20.04 starts using systemd FINALLY!!! 
2. Ubuntu 20.04 via wsl -d Ubuntu 20.04 and WT - open new tab and switch to Ubuntu 20.04 - from /init, as in the past
3 . My CentOS7 - like Ubuntu 20.04
............................................................................... 
Depends on the Distro and startup method. 
Obviously, starting using systemd is prefered method because many Linux service implementations today are implemented to use systemd: Docker, Podman, Jornal, DBus service

**Comment by dmchitnis** [comment]
when I run wsl -l -v I get the following, which suggests that Ubuntu-20.04 is running as required. The issue is VS Code does not recognize this and says 'No WSL distros found'

  NAME            STATE           VERSION
* Ubuntu-20.04    Running         2

**Comment by bamurtaugh** [comment]
Just to verify, do you see the same distro when you run `wslconfig /list`? And does trying any of the steps [here](https://stackoverflow.com/questions/64108941/visual-studio-code-not-running-working-on-wsl-2) help?

**Comment by dmchitnis** [comment]
There is no .vscode-server in my user folder in WSL distro.

![image](https://user-images.githubusercontent.com/55762818/98026301-e9590980-1dd8-11eb-8bb6-a3c3e437de0d.png)

When I run wslconfig /list in command prompt I get the following:

![image](https://user-images.githubusercontent.com/55762818/98026532-448afc00-1dd9-11eb-9250-b7df9fd86521.png)

**Comment by aeschli** [comment]
This is the code that we use to evaluate the distros. Can you run it with node?
```
const cp = require('child_process');

const is32ProcessOn64Windows = process.env.hasOwnProperty('PROCESSOR_ARCHITEW6432');
const systemRoot = process.env['SystemRoot'];
const wslConfigPath = path.join(systemRoot, is32ProcessOn64Windows ? 'Sysnative' : 'System32', 'wslconfig.exe');

const commandArgs = ['/list'];
cp.execFile(wslConfigPath, commandArgs, { windowsVerbatimArguments: true, encoding: 'utf16le' }, (error, stdout, _) => {
    if (error) {
        console.error(error);
    } else {
        const endl = /\r\n/.test(stdout) ? '\r\n' : '\n';
        const lines = stdout.split(endl);
        let distros = [];
        for (let i = 1; i < lines.length; i++) {
            let line = lines[i];
            let m1 = line.match(/^[a-zA-Z0-9\.\-_]+/);
            if (m1) {
                distros.push(m1[0]);
            }
        }
        console.log(distros.join('\n'));
    }
});
```

**Comment by dmchitnis** [comment]
Had to add the line 'var path = require("path")' at the top. Executed that piece of code and what is returned is:
Ubuntu-20.04

**Comment by aeschli** [comment]
I don't have an explanation then why the remote extension can not find any distros. Any messages in the logs? Please check out the WSL log (`Remote-WSL: Show Log`)

**Comment by dmchitnis** [comment]
I don't see an option for Remo-WSL: Show Log in Command Palette. How do I enable it?

**Comment by aeschli** [comment]
It's always there. If you have `Remote-WSL: New Window` you also have `Remote-WSL: Show Log`. Can you check again?

**Comment by dmchitnis** [comment]
Ok, found it. I had to directly go to Command Palette instead of clicking on the extension provided link in bottom left. Here is the screenshot. Looks like the execution of wslconfig.exe is failing.

![image](https://user-images.githubusercontent.com/55762818/98385698-447b3e00-201d-11eb-9787-1392fff04fe3.png)

**Comment by dmchitnis** [comment]
When I do run this in a command prompt, it seems to work.

![image](https://user-images.githubusercontent.com/55762818/98385957-9cb24000-201d-11eb-9139-1d64cf9d8194.png)

**Comment by dmchitnis** [comment]
It looks like from within VS Code there are no permissions to execute wslconfig.exe. See below. Also note that VS Code is actually running as Administrator.

![image](https://user-images.githubusercontent.com/55762818/98994228-7862d180-24fd-11eb-8fca-d92e59c76e98.png)

**Comment by aeschli** [comment]
But then the error is correct, no WSL and distros available for the given user (Administrator).

Why are you running VSCode as Administrator?

**Comment by dmchitnis** [comment]
I'm not sure why it is running as Administrator, because I do not 'Run as Administrator' and the shortcut is not setup to run as administrator.
![image](https://user-images.githubusercontent.com/55762818/99114794-65b0d100-25bf-11eb-9f02-362df25efbf7.png)

**Comment by dmchitnis** [comment]
Uninstalled 'System' install and then did a 'User' install. VS Code still reports userid as the logged on user's id and displays [Administrator] on the title. It appears that VS Code is merely reporting that the logged on user is part of the local 'Administrator' group. When the C:\WINDOWS\System32\wslconfig.exe /list command is executed in the VS Code terminal (CMD) I still get Access Denied. So it sounds like a permissions issue. The 'terminal' obviously inherits the permissions of the user account under which VS Code is executing. It is strange however, that the logged on user is local administrator, yet VS Code terminal says 'Access Denied'.

**Comment by dmchitnis** [comment]
@aeschli what are your thoughts based on all of the previous inputs? Is there any other info that you need from me? Happy to help.

**Comment by dmchitnis** [comment]
Here is another error, not WSL, but may be related:

Internal Error - System.UnauthorizedAccessException: Access to the path 'c:\Users\dchitnis\Application Data' is denied.
   at System.IO.Enumeration.FileSystemEnumerator`1.CreateRelativeDirectoryHandle(ReadOnlySpan`1 relativePath, String fullPath)
   at System.IO.Enumeration.FileSystemEnumerator`1.MoveNext()
   at System.Collections.Generic.LargeArrayBuilder`1.AddRange(IEnumerable`1 items)
   at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at System.IO.Directory.GetFiles(String path, String searchPattern, SearchOption searchOption)
   at Microsoft.AspNetCore.Razor.LanguageServer.ProjectConfigurationFileChangeDetector.GetExistingConfigurationFiles(String workspaceDirectory) in /_/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/ProjectConfigurationFileChangeDetector.cs:line 119
   at Microsoft.AspNetCore.Razor.LanguageServer.ProjectConfigurationFileChangeDetector.StartAsync(String workspaceDirectory, CancellationToken cancellationToken) in /_/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/ProjectConfigurationFileChangeDetector.cs:line 56
   at Microsoft.AspNetCore.Razor.LanguageServer.RazorFileChangeDetectorManager.InitializedAsync() in /_/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorFileChangeDetectorManager.cs:line 46
   at Microsoft.AspNetCore.Razor.LanguageServer.RazorLanguageServer.<>c__DisplayClass9_0.<<CreateAsync>b__2>d.MoveNext() in /_/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorLanguageServer.cs:line 99
--- End of stack trace from previous location ---
   at OmniSharp.Extensions.LanguageServer.Server.LanguageServer.MediatR.IRequestHandler<OmniSharp.Extensions.LanguageServer.Protocol.Models.InternalInitializeParams,OmniSharp.Extensions.LanguageServer.Protocol.Models.InitializeResult>.Handle(InternalInitializeParams request, CancellationToken token)
   at OmniSharp.Extensions.LanguageServer.Server.Pipelines.SemanticTokensDeltaPipeline`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
   at OmniSharp.Extensions.LanguageServer.Server.Pipelines.ResolveCommandPipeline`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
   at MediatR.Pipeline.RequestPreProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
   at MediatR.Pipeline.RequestPostProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
   at MediatR.Pipeline.RequestExceptionProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
   at MediatR.Pipeline.RequestExceptionProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
   at MediatR.Pipeline.RequestExceptionActionProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
   at MediatR.Pipeline.RequestExceptionActionProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
   at OmniSharp.Extensions.JsonRpc.RequestRouterBase`1.<RouteRequest>g__InnerRoute|5_0(IServiceScopeFactory serviceScopeFactory, Request request, TDescriptor descriptor, Object params, CancellationToken token, ILogger logger)
   at OmniSharp.Extensions.JsonRpc.RequestRouterBase`1.RouteRequest(IRequestDescriptor`1 descriptors, Request request, CancellationToken token)
   at OmniSharp.Extensions.JsonRpc.InputHandler.<>c__DisplayClass38_0.<<RouteRequest>b__5>d.MoveNext()

**Comment by dmchitnis** [comment]
@bamurtaugh , @aeschli any update on this issue? Is there any other help you need from me to resolve this issue?

**Comment by aeschli** [comment]
@dmchitnis Sorry I have no idea. What about uninstalling VSCode and install it as a User install?

**Comment by dmchitnis** [comment]
> Uninstalled 'System' install and then did a 'User' install. VS Code still reports userid as the logged on user's id and displays [Administrator] on the title. It appears that VS Code is merely reporting that the logged on user is part of the local 'Administrator' group. When the C:\WINDOWS\System32\wslconfig.exe /list command is executed in the VS Code terminal (CMD) I still get Access Denied. So it sounds like a permissions issue. The 'terminal' obviously inherits the permissions of the user account under which VS Code is executing. It is strange however, that the logged on user is local administrator, yet VS Code terminal says 'Access Denied'.

I have tried that, please see above.

**Comment by MedAnd** [comment]
Howdy... I'm having a very similar issue using latest VS Code and trying to launch remote using WSL2 and Ubuntu-20.04. The error log is reproduced below:

> Resolving wsl+Ubuntu-20.04, resolveAttempt: 1
Starting VS Code Server inside WSL (Ubuntu-20.04)
Extension version: 0.52.0, Windows build: 19042. Multi distro support: available. WSL path support: enabled
No shell environment set or found for current distro.
Probing if server is already installed: C:\WINDOWS\System32\wsl.exe -d Ubuntu-20.04 -e sh -c "[ -d ~/.vscode-server/bin/ea3859d4ba2f3e577a159bc91e3074c5d85c0523 ] && printf found || ([ -f /etc/alpine-release ] && printf alpine-; uname -m)"
Probing result: x86_64
No server install found in WSL, needs x64
Launching C:\WINDOWS\System32\wsl.exe -d Ubuntu-20.04 sh -c '"$VSCODE_WSL_EXT_LOCATION/scripts/wslServer.sh" ea3859d4ba2f3e577a159bc91e3074c5d85c0523 stable .vscode-server 0  '}
sh: 1: /mnt/c/Users/Admin/.vscode/extensions/ms-vscode-remote.remote-wsl-0.52.0/scripts/wslServer.sh: Permission denied
VS Code Server for WSL closed unexpectedly.
For help with startup problems, go to
https://code.visualstudio.com/docs/remote/troubleshooting#_wsl-tips
WSL Daemon exited with code 0

**Comment by dmchitnis** [comment]
It appears that 'child_process' may not be the right way to get this information particularly because system32 may be a protected folder along with 'program files' and 'program files(x86)' and thereby being blocked. Unblocking is not the solution as this may render the system vulnerable to malicious attacks. There has to be a better way to accomplish this. From the comment above it seems a similar approach is being taken to execute wslServer.sh via wsl.exe landing up in Permission denied.

**Comment by MedAnd** [comment]
Hi @aeschli @bamurtaugh - any update on this bug? It's preventing adoption of VS Code and WSL2 for Kubernetes development. 

Thanks.

**Comment by aeschli** [comment]
@MedAnd Are you sure your problem is the same as the original issue ? If not, would you mind filing a separate issue. These permission problems are puzzling enough especially when they are not reproducible on my machine.

Here's what @dmchitnis has reported (@dmchitnis please correct me if I got this wrong)
- open a Windows command prompt, run `C:\WINDOWS\System32\wslconfig.exe /list` correctly lists the available distros
- open VSCode in Windows, in the integrated terminal run `C:\WINDOWS\System32\wslconfig.exe /list`: Access denied
- The VSCode title bars says `[Administrator]`

`[Administrator]` indicates that VSCode is run in elevated mode You could verify that information in the Task manager by going to the Details view and adding `Elevated` as a column. 
When I run elevated I get the Windows elevated prompt first.

Can you run the command prompt or powershell as admin and see if `C:\WINDOWS\System32\wslconfig.exe /list`  works?

**Comment by dmchitnis** [comment]
It does look like VS Code runs in elevated mode (although I never get a UAC prompt, which normally I do for other applications). I checked the C:\WINDOWS\System32\wslconfig.exe /list command in an elevated command prompt and I did get access denied. I checked all properties of shortcuts and also the Code.exe to make sure none of the 'Run as Administrator' boxes are checked. So wonder why VS Code runs in elevated mode.

With the fact established that C:\WINDOWS\System32\wslconfig.exe /list is not allowed in elevated mode, what next?

**Comment by ednl** [comment]
Not exactly the same problem, maybe, or maybe it is: Code does not run as Administrator (it's a user install) but I do get the exact same error:

```
[2021-01-24 18:49:19.680] No server install found in WSL, needs x64
[2021-01-24 18:49:19.682] Launching C:\WINDOWS\System32\wsl.exe -d Ubuntu-20.04 sh -c '"$VSCODE_WSL_EXT_LOCATION/scripts/wslServer.sh" ea3859d4ba2f3e577a159bc91e3074c5d85c0523 stable .vscode-server 0  --disable-telemetry'}
[2021-01-24 18:49:19.928] sh: 1: /mnt/c/Users/myusername/.vscode/extensions/ms-vscode-remote.remote-wsl-0.52.0/scripts/wslServer.sh: Permission denied
[2021-01-24 18:49:19.928] VS Code Server for WSL closed unexpectedly.
```

Same for Ubuntu-20.04 and Debian, which both run fine in Windows Terminal. I tried deleting the C:\Users\myusername\.vscode directory and re-installing Remote-WSL, but the error stays the same. I can't use VS Code and WSL together anymore!

I wonder what exactly throws the permission error: is it Windows or Linux inside WSL? I do have my home directory in WSL set as chmod 700.

**Comment by JasonMiller-BAH** [comment]
> @MedAnd Are you sure your problem is the same as the original issue ? If not, would you mind filing a separate issue? These permission problems are puzzling enough especially when they are not reproducible on my machine.
> 
> Here's what @dmchitnis has reported (@dmchitnis please correct me if I got this wrong)
> 
> * open a Windows command prompt, run `C:\WINDOWS\System32\wslconfig.exe /list` correctly lists the available distros
> * open VSCode in Windows, in the integrated terminal run `C:\WINDOWS\System32\wslconfig.exe /list`: Access denied
> * The VSCode title bars says `[Administrator]`
> 
> `[Administrator]` indicates that VSCode is run in elevated mode You could verify that information in the Task manager by going to the Details view and adding `Elevated` as a column.
> When I run elevated I get the Windows elevated prompt first.
> 
> Can you run the command prompt or powershell as admin and see if `C:\WINDOWS\System32\wslconfig.exe /list` works?

Chiming to say that I'm getting exactly the same errors as @dmchitnis, which are aptly summarized here.

> It does look like VS Code runs in elevated mode (although I never get a UAC prompt, which normally I do for other applications). I checked the C:\WINDOWS\System32\wslconfig.exe /list command in an elevated command prompt and I did get access denied. I checked all properties of shortcuts and also the Code.exe to make sure none of the 'Run as Administrator' boxes are checked. So wonder why VS Code runs in elevated mode.
> 
> With the fact established that C:\WINDOWS\System32\wslconfig.exe /list is not allowed in elevated mode, what next? And another question less important is how do I stop VS Code from running in elevated mode? If for some reason VS Code is made to run in elevated mode by a policy, I may not have control over running it as a regular user.
> 
> Here is a link for another user having a similar issue, so it seems quite widespread https://stackoverflow.com/questions/54904572/vs-code-always-starting-in-administrator-mode

Confirmed.  My VSCode always runs elevated but without ever prompting me for elevated permissions.  An elevated command prompt or PowerShell results in the expected denial of access to `wslconfig.exe /list`.

**Comment by JasonMiller-BAH** [comment]
In the [other thread](https://github.com/microsoft/vscode/issues/115715), users have narrowed this down to a problem with BeyondTrust.  [One user](https://github.com/microsoft/vscode/issues/115715#issuecomment-782239147) reports that his IT folks managed a workaround and will get back to us with details when possible.

**Comment by dmchitnis** [comment]
I can confirm that I do have BeyondTrust Privilege Management running on my computer. However, I have no control over it.

**Comment by rainabba** [comment]
I do not believe I'm dealing with the same cause, but appear to have the same issue. I updated my original 18.04 distro internally, then exported and imported as 20.04 (not a rename, but a clean import) then removed the old distro. Now I'm launching vscode from my Ubuntu 20.04 session using Terminal, then launching `code &` and getting `Unknown distro: ubuntu18.04` from the Remote - WSL Extension.

 I grepped for '18.04' and do find it in the following. Do I need to edit manually to get things working again, try a reinstall or is there a more elegant/simple solution to get VSCode up to speed on the new distro name? I can't even File -> Open in vscode right now because of this. All else is fine though, including starting code from WSL. I would expect to find a setting in WSL - Remote extension, but don't see anything there.

``` bash
PS C:\Users\me> wslconfig /list
Windows Subsystem for Linux Distributions:
Ubuntu-20.04 (Default)
```


RESOLUTION: I was launching code from the imported wsl instance and it was trying to use a filepath to the old \\wsl$\\instance. It turns out I only needed to use File -> Close Remote Connection,[ maybe not needed, but I resinstalled the "WSL - Remote" extension, ] then connect to WSL using the green button at the bottom-left (or File -> Reopen in WSL.)
